crypto/engine/eng_ctrl.c

```diff
@@ -124,19 +124,13 @@ static int int_ctrl_helper(ENGINE *e, int cmd, long i, void *p,
 
 int ENGINE_ctrl(ENGINE *e, int cmd, long i, void *p, void (*f) (void))
 {
-    int ctrl_exists, ref_exists;
+    int ctrl_exists;
+
     if (e == NULL) {
         ENGINEerr(ENGINE_F_ENGINE_CTRL, ERR_R_PASSED_NULL_PARAMETER);
         return 0;
     }
-    CRYPTO_THREAD_write_lock(global_engine_lock);
-    ref_exists = ((e->struct_ref > 0) ? 1 : 0);
-    CRYPTO_THREAD_unlock(global_engine_lock);
-    ctrl_exists = ((e->ctrl == NULL) ? 0 : 1);
-    if (!ref_exists) {
-        ENGINEerr(ENGINE_F_ENGINE_CTRL, ENGINE_R_NO_REFERENCE);
-        return 0;
-    }
+    ctrl_exists = e->ctrl != NULL;
```

Comment:
(only commenting)

Right, because this assumes that the caller already has a proper reference, i.e. the reference count *cannot* (or at least *must not*) be zero when this function is called.

Reply:
Yes.  It is impossible for the struct_ref field to be zero without the engine having been freed.  The initialisation code sets the count to one and the only decrement operation frees it when it goes to zero.

This code is redundant albeit conservative.